src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,234 @@
+			);
+
+			ctx.RegisterSymbolAction(
+				ctx => InspectMethodDeclaration(
+					ctx,
+					attr,
+					noteMethodWithBlockingAttribute: m => haveBlockingAttribute[m] = true
+				),
+				SymbolKind.Method
+			);
+
+			var callsSomethingBlocking = new ConcurrentDictionary<IMethodSymbol, bool>(
+				SymbolEqualityComparer.Default
+			);
+
+			ctx.RegisterSyntaxNodeAction(
+				ctx => InspectMethodCall(
+					ctx,
+					attr,
+					noteMethodHasBlockingCall: m => callsSomethingBlocking[m] = true
+				),
+				SyntaxKind.InvocationExpression
+			);
+
+			// After the rest of analysis report about potentially unnecessary [Blocking]s
+			ctx.RegisterCompilationEndAction(
+				ctx => WarnForUnusedBlockingAttributes(
+					ctx,
+					attr,
+					haveBlockingAttribute: haveBlockingAttribute.Keys,
+					callsSomethingBlocking: callsSomethingBlocking.Keys
+				)
+			);
+		}
+
+		private static void InspectMethodDeclaration(
+			SymbolAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodWithBlockingAttribute
+		) {
+			// Goal: find async methods that have [Blocking]
+
+			var methodSymbol = (IMethodSymbol)ctx.Symbol;
+
+			if( !methodSymbol.IsAsync ) {
+				// Make note if it has [Blocking], because it may not be necessary.
+				if( HasAttribute( methodSymbol, blockingAttr ) ) {
+					noteMethodWithBlockingAttribute( methodSymbol );
+				}
+
+				return;
+			}
+
+			var attrData = GetAttribute( methodSymbol, blockingAttr );
+
+			if( attrData == null ) {
+				return;
+			}
+
+			ctx.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.AsyncMethodCannotBeBlocking,
+					attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+					methodSymbol.Name
+				)
+			);
+
+			// TODO: code fix to remove [Blocking]
+		}
+
+		private static void InspectMethodCall(
+			SyntaxNodeAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodHasBlockingCall
+		) {
+			// Goal: find calls to blocking methods inside non-blocking methods, or in properties.
+
+			var invocation = (InvocationExpressionSyntax)ctx.Node;
+
+			var invokedSymbolInfo = ctx.SemanticModel.GetSymbolInfo( invocation.Expression );
+
+			// Unfortunately there are ways to use this to dodge analysis, but
+			// there is only so much we can do.
+			if( invokedSymbolInfo.Symbol?.Kind != SymbolKind.Method ) {
+				return;
+			}
+
+			var invokedMethodSymbol = (IMethodSymbol)invokedSymbolInfo.Symbol;
+
+			// async methods are never [Blocking] (see other parts of the analyzer)
+			if( invokedMethodSymbol.IsAsync ) {
+				return;
+			}
+
+			// If the thing we're calling isn't [Blocking] then OK.
+			// TODO: support other means of inferring blockingness, for 3rd party libraries.
+			if( !HasAttribute( invokedMethodSymbol, blockingAttr ) ) {
+				return;
+			}
+
+			var myMethodDeclaration = invocation.Ancestors()
+				.FirstOrDefault( s => s.IsKind( SyntaxKind.MethodDeclaration ) )
+				as MethodDeclarationSyntax;

```

Comment:
```suggestion
			var myMethodDeclaration = invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>();
```

Reply:
Also, might need to think about lamdas/anonymous functions

Reply:
Made it more robust